src/api/app/models/worker_status.rb

```diff
@@ -1,25 +1,45 @@
 class WorkerStatus
-  def self.hidden
-    mydata = Rails.cache.read('workerstatus')
-    ws = Nokogiri::XML(mydata || Backend::Api::BuildResults::Worker.status).root
-    prjs = {}
-    ws.css('building').each do |b|
-      prjs[b['project']] = 1
+  class << self
+    def hidden
+      mydata = Rails.cache.fetch('workerstatus') do
+        Backend::Api::BuildResults::Worker.status
+      end
+      # remove information about projects which are not visible to the current user
+      hide_project_information(Nokogiri::XML(mydata).root)
+    end
+
+    private
+
+    def hide_project_information(worker_status_root)
+      worker_status = worker_status_root
+      prjs = initialize_projects(worker_status)
+      xpath_string = generate_xpath(project_names(prjs.keys))
+      worker_status.xpath(xpath_string).each do |b|
+        b['project'] = '---'
+        b['repository'] = '---'
+        b['package'] = '---'
+      end
+      worker_status
     end
-    names = {}
-    # now try to find those we have a match for (the rest are hidden from you
-    Project.where(name: prjs.keys).pluck(:name).each do |n|
-      names[n] = 1
+
+    # use Xpath to filter the projects
+    def generate_xpath(names)
+      return '//building' if names.empty?
+      "//building[not(@project=#{xpath_list(names)})]"
```

Comment:
We try to find all `building` entries where the project doesn't meet one of the elements of the list

http://saxon.sourceforge.net/saxon7.9/expressions.html#sequence